### ISSUE TITLE
Unable to get GPU working

### ISSUE THREAD

**Issue by jaredallard** [chadmed/asahi-overlay#51]
Now I promise I'm not hitting the usual "forgot to compile with options (or rust isn't setup)" issue that usually is the cause of this.

I can't get the GPU to work with Mesa 24 and the current 6.6.0 kernel :( Hopeful someone will have an idea of what I might be missing. From all of the information below, I'm lead to believe it's a mesa issue or something else related to that that I'm missing. I've done `asahi-fwupdate` and `update-m1n1` recently.

### Information

[**Kernel ebuild**](https://github.com/jaredallard/asahi-overlay/blob/main/sys-kernel/asahi-kernel/asahi-kernel-6.6.0_p6.ebuild)

**`uname -a`**: `Linux yumi 6.6.0-p6-asahi-dist #1 SMP PREEMPT_DYNAMIC Tue Dec  5 14:59:48 PST 2023 aarch64 GNU/Linux`

**Kernel Config (`/proc/config.gz`)**:

```bash
CONFIG_RUST_IS_AVAILABLE=y
CONFIG_RUST=y
CONFIG_RUSTC_VERSION_TEXT="rustc 1.73.0 (cc66ad468 2023-10-03) (gentoo)"
CONFIG_HAVE_RUST=y
CONFIG_RUST_DRM_SCHED=y
CONFIG_RUST_DRM_GEM_SHMEM_HELPER=y
CONFIG_RUST_APPLE_RTKIT=y
CONFIG_DRM_ASAHI=m
# CONFIG_DRM_ASAHI_DEBUG_ALLOCATOR is not set
CONFIG_DRM_APPLE=m
# CONFIG_RUST_DEBUG_ASSERTIONS is not set
CONFIG_RUST_OVERFLOW_CHECKS=y
# CONFIG_RUST_BUILD_ASSERT_ALLOW is not set
```

**Mesa Version**: `24.0.0_pre20231126` (**UPDATE**: Even Mesa 23 doesn't work... so I guess it's related to the Kernel somehow :( )

`dmesg | grep asahi`:

```bash
[   12.577512] asahi 406400000.gpu: Probing...
[   12.578324] asahi 406400000.gpu: MMU: Initializing...
[   12.579914] asahi 406400000.gpu: MMU: Initializing kernel page table
[   12.580333] asahi 406400000.gpu: MMU: Creating kernel page tables
[   12.580690] asahi 406400000.gpu: MMU: Kernel page tables created
[   12.581025] asahi 406400000.gpu: MMU: Initializing TTBs
[   12.581355] asahi 406400000.gpu: MMU: initialized
[   12.581665] asahi 406400000.gpu: GPU ID registers: 0x4032000 0x40a06 0x30808 0x40404 0x1320200 0x202311
[   12.582004] asahi 406400000.gpu: Core masks: [
[   12.584744] asahi 406400000.gpu: GPU Information:
[   12.585149] asahi 406400000.gpu:   Type: G13S
[   12.585151] asahi 406400000.gpu:   Clusters: 2
[   12.585152] asahi 406400000.gpu:   Cores: 8 (16)
[   12.585152] asahi 406400000.gpu:   Frags: 8 (16)
[   12.585153] asahi 406400000.gpu:   GPs: 4 (8)
[   12.585153] asahi 406400000.gpu:   Core masks: [
[   12.589461] asahi 406400000.gpu:   Active cores: 14
[   12.589462] asahi 406400000.gpu: Getting configuration from device tree...
[   12.589482] asahi 406400000.gpu: Dynamic configuration fetched
[   12.591904] asahi 406400000.gpu: MMU: IO map: 0x404d00000:0x1c000 -> 0xffffffaf00000000
[   12.591906] asahi 406400000.gpu: MMU: IO map: 0x20e100000:0x4000 -> 0xffffffaf00020000
[   12.591907] asahi 406400000.gpu: MMU: IO map: 0x28e104000:0x4000 -> 0xffffffaf00028000
[   12.591908] asahi 406400000.gpu: MMU: IO map: 0x404000000:0x20000 -> 0xffffffaf00030000
[   12.591909] asahi 406400000.gpu: MMU: IO map: 0x28e494000:0x4000 -> 0xffffffaf00054000
[   12.591910] asahi 406400000.gpu: MMU: IO map: 0x404d80000:0x8000 -> 0xffffffaf0005c000
[   12.591911] asahi 406400000.gpu: MMU: IO map: 0x204d60000:0x4000 -> 0xffffffaf00068000
[   12.591911] asahi 406400000.gpu: MMU: IO map: 0x200000000:0xd8000 -> 0xffffffaf00070000
[   12.591912] asahi 406400000.gpu: MMU: IO map: 0x201000000:0xd8000 -> 0xffffffaf00148000
[   12.591913] asahi 406400000.gpu: MMU: IO map: 0x202000000:0xd8000 -> 0xffffffaf00220000
[   12.591914] asahi 406400000.gpu: MMU: IO map: 0x203000000:0xd8000 -> 0xffffffaf002f8000
[   12.591915] asahi 406400000.gpu: MMU: IO map: 0x2643c4000:0x4000 -> 0xffffffaf003d4000
[   12.591915] asahi 406400000.gpu: MMU: IO map: 0x28e3d0000:0x4000 -> 0xffffffaf003dc000
[   12.591916] asahi 406400000.gpu: MMU: IO map: 0x28e3c0000:0x4000 -> 0xffffffaf003e4000
[   12.593729] asahi 406400000.gpu: RTKit: Initializing (protocol version 12)
[   12.601192] [drm] Initialized asahi 0.0.0 20220831 for 406400000.gpu on minor 1
[   12.601856] asahi 406400000.gpu: Probed!
```

**Comment by chadmed** [comment]
silly question, do you have `VIDEO_CARDS="asahi"`  in your `make.conf`? Your kernel looks fine, apart from those DCP errors which is indicative of the crossbar driver missing.

**Comment by jaredallard** [comment]
> silly question, do you have `VIDEO_CARDS="asahi"` in your `make.conf`? Your kernel looks fine, apart from those DCP errors which is indicative of the crossbar driver missing.

I'm about 99.9% sure I do, but I'll double check (unfortunately cannot access my personal computer until end of day PT 😢)

**Comment by jaredallard** [comment]
I've been using this machine as a bin host ([mesa gpkg link, if helpful at all](https://gentoo.rgst.io/t/arm64/m1/media-libs/mesa/mesa-24.0.0_pre20231120-1.gpkg.tar)), so I was able to grab information about my packages from that (since `gpkg` contains that information). From the `USE`, it would seem like I do:

```
X arm64 elibc_glibc gles2 kernel_linux llvm proprietary-codecs video_cards_asahi vulkan wayland zstd
```

**Comment by enx1998** [comment]
Hi, I have the same issue, but I don't have `apple-dcp 289c00000.dcp: Failed to get dp-xbar: -517` error message in dmesg. With the kernel version 6.5.0_p29 everything works as expected.
uname -a: `
Linux gentoom1 6.6.0-asahi-6-ARCH #1 SMP PREEMPT_DYNAMIC Fri Dec  8 02:05:55 CET 2023 aarch64 GNU/Linux`

**Comment by jaredallard** [comment]
Adding the config option did remove the `apple-dcp` line. I tried using `6.5.0_p29` but that didn't seem to change anything (for me).

**Comment by chadmed** [comment]
can you provide the output of `asahi-diagnose`?

**Comment by jaredallard** [comment]
I was on the macOS beta and had upgraded the firmware around the time this happened, so I decided to do a full reset to the stable version of macOS and do a complete reinstall of my Gentoo install to test that process again (this time through Fedora, will have some code changes coming in a PR to gentoosupport soon for that). So, I'll see if that fixes this up, if not I'll give that a run (just... might be a hot second :D)

**Comment by enx1998** [comment]
I have found what's the problem: my case with kernel 6.6 the asahi gpu driver was not compiled because I had a version of `dev-util/bindgen` too old (installed 0.62.0, needed at least 0.65), so I installed the version `0.68.1`  available in portage and now the `asahi` gpu driver has compiled and everything is working correctly.
@chadmed I noticed that the HDMI doesn't work (M1 pro 14" with Sonoma 14.2.1), it is beacause when I installed Gentoo I had an old version of MacOS? Do I have to reinstall Gentoo to get the HDMI to work?

Thanks ENx

**Comment by jaredallard** [comment]
@enx1998 I think you need to boot back into macOS and re-run the install script. It'll have the option (when you would normally resize to install a new one) to update the firmware, which should do it (or might end up in my situation, 50/50 :D)

**Comment by jaredallard** [comment]
Looks like that didn't help, but at least I got to have fun doing that again. Here's the generated text output of `asahi-diagnose`:

[asahi-diagnose-20231209-195625.txt](https://pastebin.com/Un8RAqn1)

**Comment by chadmed** [comment]
hm, your gpu is probing fine. can you try with the latest mesa and kernel? its possible your kernel and mesa went out of sync

**Comment by jaredallard** [comment]
Still digging into this, noting what I've found so far from trying to debug this. The `dri` devices seem to be created fine and appear to be valid:

```bash
$ ls -l /dev/dri
total 0
drwxr-xr-x  2 root root        100 Dec 22 16:11 by-path
crw-rw----+ 1 root video  226,   0 Dec 22 16:11 card0
crw-rw----+ 1 root video  226,   1 Dec 22 16:11 card1
crw-rw-rw-  1 root render 226, 128 Dec 22 16:11 renderD128

$ ls -l /sys/class/drm/renderD*/device/driver | awk '{print $9,$11}'
/sys/class/drm/renderD128/device/driver ../../../../bus/platform/drivers/asahi

$ ls -alh /sys/class/drm/renderD128/device
lrwxrwxrwx 1 root root 0 Dec 22 16:12 /sys/class/drm/renderD128/device -> ../../../406400000.gpu

$ ls -alh /sys/class/drm/card1/device
lrwxrwxrwx 1 root root 0 Dec 22 16:12 /sys/class/drm/card1/device -> ../../../406400000.gpu
```

But, trying to run Plasma doesn't use it (`glxinfo -B`):

```bash
glxinfo -B
name of display: :0
MESA-LOADER: failed to open zink: /usr/lib64/dri/zink_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib64/dri, suffix _dri)
failed to load driver: zink
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa (0xffffffff)
    Device: llvmpipe (LLVM 17.0.6, 128 bits) (0xffffffff)
    Version: 24.0.0
    Accelerated: no
    Video memory: 15589MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 4.5
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
Memory info (GL_ATI_meminfo):
    VBO free memory - total: 63 MB, largest block: 63 MB
    VBO free aux. memory - total: 13534 MB, largest block: 13534 MB
    Texture free memory - total: 63 MB, largest block: 63 MB
    Texture free aux. memory - total: 13534 MB, largest block: 13534 MB
    Renderbuffer free memory - total: 63 MB, largest block: 63 MB
    Renderbuffer free aux. memory - total: 13534 MB, largest block: 13534 MB
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 4294088535 MB
    Total available memory: 4294104124 MB
    Currently available dedicated video memory: 63 MB
OpenGL vendor string: Mesa
OpenGL renderer string: llvmpipe (LLVM 17.0.6, 128 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 24.0.0-devel
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.5 (Compatibility Profile) Mesa 24.0.0-devel
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 24.0.0-devel
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

I don't have a ton of experience debugging these issues and couldn't find a way to force mesa to use a specific `dri` device, but I _could_ find a way to get `wlroots` to do so. So, I used `Hyprland` to try to force it to use `card1`, which is the asahi GPU. This also fails, but gives us a little bit of information:

```bash
[LOG] Disabling stdout logs! Check the log for further logs.
[wlr] [backend/session/session.c:410] Unable to open /dev/dri/card1 as DRM device
[wlr] [backend/backend.c:210] Found 0 GPUs, cannot create backend
[wlr] [backend/backend.c:376] Failed to open any DRM device
[CRITICAL] m_sWLRBackend was NULL!
```

Just in case, I tried the `render` device as well:

```
[wlr] [libseat] [libseat/backend/logind.c:137] Could not take device: No such device
[wlr] [backend/session/session.c:321] Failed to open device: '/dev/dri/renderD128': Resource temporarily unavailable
[wlr] [backend/session/session.c:410] Unable to open /dev/dri/renderD128 as DRM device
[wlr] [backend/backend.c:210] Found 0 GPUs, cannot create backend
[wlr] [backend/backend.c:376] Failed to open any DRM device
```

**Comment by jaredallard** [comment]
Well. Somehow, during my reinstall, `CONFIG_DRM_APPLE` got unset. Setting that fixed it (as would be expected...). I don't know what caused the original scenario, but this is fixed.
